### PR TITLE
Add support for EAS Update

### PR DESCRIPTION
**Pull request by kbrandwijk**
# Checklist

- [ ] I've added an entry to [CHANGELOG.md](https://github.com/expo/sentry-expo/blob/main/CHANGELOG.md#master) if necessary.

# Why

This PR adds support for EAS Update.

# How

This PR adds support for EAS Update through the following set of changes:
- Uses the new `Updates.isEmbeddedLaunch` property of `expo-updates` to determine whether a build or an update is loaded.
- For the base build, uses the existing logic for setting the `dist` value that matches the default behavior of the embedded scripts that are used at build time to set `dist`
- For updates, uses the `Updates.updateId` as unique identifier for `dist`. The `release` field is untouched, as the update applies to the same release (major version and build number/version code).
- Upgrading `expo-updates` required an upgrade to Expo SDK 47. This will be the minimum supported SDK release for this version.
- As part of the upgrade, `@sentry/react-native` is also bumped to the latest version
- Sentry has introduced `@sentry/react` as a wrapper for `@sentry/browser`. As this is more applicable to the web projects this libary is used for, this is now exported as `Sentry.React.*`. The exports also stay available through `Sentry.Browser.*` for backwards compatibility purposes.

# Test Plan

This version of `sentry-expo` has been successfully used in manual testing in an app for both builds and updates.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,18 +37,18 @@
...
-    "@expo/spawn-async": "^1.6.0",
-    "@sentry/browser": "7.7.0",
-    "@sentry/integrations": "7.7.0",
-    "@sentry/react-native": "4.2.2",
-    "@sentry/types": "7.7.0",
+    "@expo/config-plugins": "~5.0.4",
+    "@expo/config-types": "^47.0.0",
+    "@expo/spawn-async": "^1.7.0",
+    "@sentry/integrations": "7.16.0",
+    "@sentry/react": "^7.21.0",
```
Comment (SimenB) [inline package.json:44]: this should match the versions of the other sentry packages

Comment (kbrandwijk) [reply package.json]: solved

File: package.json [hunk 2]
```diff
@@ -37,11 +37,11 @@
...
   "dependencies": {
-    "@expo/spawn-async": "^1.6.0",
-    "@sentry/integrations": "7.16.0",
-    "@sentry/react": "7.16.0",
-    "@sentry/react-native": "4.8.0",
-    "@sentry/types": "7.16.0",
+    "@expo/spawn-async": "^1.7.0",
+    "@sentry/integrations": "7.20.1",
+    "@sentry/react": "7.20.1",
+    "@sentry/react-native": "4.9.0",
```
Comment (SimenB) [inline package.json:43]: haha, great timing 😀

File: src/sentry.ts [hunk 3]
```diff
@@ -20,7 +20,7 @@ const IS_BARE_WORKFLOW = Constants.executionEnvironment === ExecutionEnvironment
 const DEFAULT_OPTIONS = {
   enableNativeNagger: false, // Otherwise this will trigger an Alert(), let's rely on the logs instead
   release: getDefaultRelease(),
-  dist: MANIFEST.revisionId ? MANIFEST.version : `${Application.nativeBuildVersion}`,
+  dist: Updates.isEmbeddedLaunch ? (MANIFEST.revisionId ? MANIFEST.version : `${Application.nativeBuildVersion}`) : Updates.updateId,
```
Comment (brentvatne) [inline src/sentry.ts:23]: nit: i'd move this conditional out into a separate function so it's easier to read

**Review by brentvatne — APPROVED** [review]

**Comment by SimenB** [comment]
🎉 

Does https://github.com/expo/expo/blob/4fdd6db265db5fac025fffd1dbc33143d7c8a7ed/packages/expo/bundledNativeModules.json#L103 and https://github.com/expo/expo/blob/4fdd6db265db5fac025fffd1dbc33143d7c8a7ed/packages/expo/bundledNativeModules.json#L108 need to be updated?

**Comment by kbrandwijk** [comment]
> 🎉
> 
> Does https://github.com/expo/expo/blob/4fdd6db265db5fac025fffd1dbc33143d7c8a7ed/packages/expo/bundledNativeModules.json#L103 and https://github.com/expo/expo/blob/4fdd6db265db5fac025fffd1dbc33143d7c8a7ed/packages/expo/bundledNativeModules.json#L108 need to be updated?

Absolutely, which is why https://github.com/expo/expo/pull/20201 is pending merge. I always want to make sure there's no issues with the release before making it the default for everyone.

**Comment by solarisn** [comment]
Any advice for sourcemap uploads to Sentry for EAS Updates? Will there be support for automatic upload as a post-publish hook when deploying an EAS Update?

**Comment by SimenB** [comment]
You can try to adapt https://github.com/expo/expo/pull/19753#issuecomment-1350969824

**Comment by solarisn** [comment]
There are new instructions for managing sourcemap upload when running using EAS Updates inside of the official Expo documentation:
[https://docs.expo.dev/guides/using-sentry/#uploading-source-maps-for-updates](https://docs.expo.dev/guides/using-sentry/#uploading-source-maps-for-updates)

**Comment by nandorojo** [comment]
> There are new instructions for managing sourcemap upload when running using EAS Updates inside of the official Expo documentation: https://docs.expo.dev/guides/using-sentry/#uploading-source-maps-for-updates

Just to give my input here...I found these instructions a bit confusing. Thought I'd share in case it's helpful. 

I'm not familiar with parsing the EAS output and running a script based on it. The typical "Expo-y" way would be to handle this at the config plugin / postPublish step without needing to add this to CI manually.

Would be awesome for that to be the case! But for now, I'll try implementing this comment: https://github.com/expo/expo/pull/19753#issuecomment-1350969824

At the very least, I think the docs should link to that example. Thanks for all the improvements here though!

**Pull request merged**